Compiling `span_tests.cpp` fails on Linux with `clang++` using `libstdc++`

Creating a `std::match_result<BiIter>` fails (line 1530) when using a `gsl::span<char>::iterator`.

**error: exception specification of explicitly defaulted move constructor does not match the calculated one**

```cpp
    std::match_results<span<char>::iterator> match; // ERROR!!

    std::regex_match(s.begin(), s.end(), match, std::regex(".*"));
    CHECK(match.ready());
    CHECK(!match.empty());
```

```
In file included from /home/galik/tmp/GSL/tests/span_tests.cpp:25:
In file included from /usr/bin/../lib/gcc/x86_64-redhat-linux/6.3.1/../../../../include/c++/6.3.1/regex:62:
/usr/bin/../lib/gcc/x86_64-redhat-linux/6.3.1/../../../../include/c++/6.3.1/bits/regex.h:1577:7: error: exception
      specification of explicitly defaulted move constructor does not match the calculated one
      match_results(match_results&& __rhs) noexcept = default;
      ^
/home/galik/tmp/GSL/tests/span_tests.cpp:1530:46: note: in instantiation of template class
      'std::__cxx11::match_results<gsl::details::span_iterator<gsl::span<char, -1>, false>,
      std::allocator<std::__cxx11::sub_match<gsl::details::span_iterator<gsl::span<char, -1>, false> > > >' requested
      here
    std::match_results<span<char>::iterator> match;
                                             ^
1 error generated.

```
Compiler version info.
```
clang version 3.9.1 (tags/RELEASE_391/final)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
```

--- a/include/gsl/span
+++ b/include/gsl/span
@@ -141,36 +141,34 @@ namespace details
         using reference = std::conditional_t<IsConst, const element_type_, element_type_>&;
         using pointer = std::add_pointer_t<reference>;
 
-        constexpr span_iterator() GSL_NOEXCEPT : span_iterator(nullptr, 0) {}
+        span_iterator() = default;
 
         constexpr span_iterator(const Span* span, typename Span::index_type index)
-            : span_(span), index_(index)
+            : span_(span), index_(index) GSL_NOEXCEPT
         {
             Expects(span == nullptr || (index_ >= 0 && index <= span_->length()));
         }
 
-        friend class span_iterator<Span, true>;
-        constexpr span_iterator(const span_iterator<Span, false>& other) GSL_NOEXCEPT
+        friend span_iterator<Span, true>;
+        template<bool B, std::enable_if_t<!B && IsConst>* = nullptr>
+        constexpr span_iterator(const span_iterator<Span, B>& other) GSL_NOEXCEPT
             : span_iterator(other.span_, other.index_)
         {
         }
 
-        constexpr span_iterator<Span, IsConst>&
-        operator=(const span_iterator<Span, IsConst>&) GSL_NOEXCEPT = default;
-
-        constexpr reference operator*() const
+        constexpr reference operator*() const GSL_NOEXCEPT
         {
             Expects(span_);
             return (*span_)[index_];
         }
 
-        constexpr pointer operator->() const
+        constexpr pointer operator->() const GSL_NOEXCEPT
         {
-            Expects(span_);
-            return &((*span_)[index_]);
+            Expects(span_ && index_ >= 0 && index_ < span_->length());
+            return span_->data() + index_;
         }
 
-        constexpr span_iterator& operator++()
+        constexpr span_iterator& operator++() GSL_NOEXCEPT
         {
             Expects(span_ && index_ >= 0 && index_ < span_->length());
             ++index_;
@@ -184,7 +182,7 @@ namespace details
             return ret;
         }
 
-        constexpr span_iterator& operator--()
+        constexpr span_iterator& operator--() GSL_NOEXCEPT
         {
             Expects(span_ && index_ > 0 && index_ <= span_->length());
             --index_;
@@ -204,7 +202,7 @@ namespace details
             return ret += n;
         }
 
-        constexpr span_iterator& operator+=(difference_type n)
+        constexpr span_iterator& operator+=(difference_type n) GSL_NOEXCEPT
         {
             Expects(span_ && (index_ + n) >= 0 && (index_ + n) <= span_->length());
             index_ += n;
@@ -219,7 +217,7 @@ namespace details
 
         constexpr span_iterator& operator-=(difference_type n) GSL_NOEXCEPT { return *this += -n; }
 
-        constexpr difference_type operator-(const span_iterator& rhs) const
+        constexpr difference_type operator-(const span_iterator& rhs) const GSL_NOEXCEPT
         {
             Expects(span_ == rhs.span_);
             return index_ - rhs.index_;
@@ -242,7 +240,8 @@ namespace details
             return !(lhs == rhs);
         }
 
-        constexpr friend bool operator<(const span_iterator& lhs, const span_iterator& rhs)
+        constexpr friend bool operator<(const span_iterator& lhs,
+                                        const span_iterator& rhs) GSL_NOEXCEPT
         {
             Expects(lhs.span_ == rhs.span_);
             return lhs.index_ < rhs.index_;
@@ -266,15 +265,9 @@ namespace details
             return !(rhs > lhs);
         }
 
-        void swap(span_iterator& rhs) GSL_NOEXCEPT
-        {
-            std::swap(index_, rhs.index_);
-            std::swap(span_, rhs.span_);
-        }
-
     protected:
-        const Span* span_;
-        std::ptrdiff_t index_;
+        const Span* span_ = nullptr;
+        std::ptrdiff_t index_ = 0;
     };
 
     template <class Span, bool IsConst>

